### PR TITLE
Adds netlink to navstar overlay

### PR DESCRIPTION
**Pull request by GoelDeepak**

### CODE REVIEW

**Comment by GoelDeepak** [comment]
@sargun ... This is ready for review

File: apps/navstar_overlay/src/navstar_overlay_netlink.erl [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+-export([ipneigh_replace/4, iproute_replace/5, bridge_fdb_replace/4,
+        iplink_show/2, iplink_add/5, iplink_set/3, iprule_show/1,
+        iprule_add/4, make_iprule/3, match_iprules/2, is_iprule_present/2,
+        ipaddr_replace/4, if_nametoindex/1]).
+
+-include_lib("gen_netlink/include/netlink.hrl").
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-define(TEST_AND_DEV, true).
```
Comment (sargun) [inline apps/navstar_overlay/src/navstar_overlay_netlink.erl]: TEST_OR_DEV?

File: apps/navstar_overlay/src/navstar_overlay_netlink.erl [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+-ifdef(DEV).
+-define(TEST_AND_DEV, true).
+-endif.
+
+%% eg. ipneigh_replace(Pid, {44,128,0,1}, {16#70,16#b3,16#d5,16#80,16#00,16#03}, "vtep1024").
+ipneigh_replace(Pid, Dst, Lladdr, Ifname) ->
+  Attr = [{dst, Dst}, {lladdr, Lladdr}],
+  Neigh = {
+    _Family = inet,
+    _Ifindex = if_nametoindex(Ifname),
```
Comment (sargun) [inline apps/navstar_overlay/src/navstar_overlay_netlink.erl]: Please put this on a separate line.

Comment (sargun) [reply apps/navstar_overlay/src/navstar_overlay_netlink.erl]: In general, if you're doing something that has side-effects while initializing a Tuple, it's a bit difficult to track down when it breaks.

**Comment by sargun** [comment]
Two things:
1) It would be good to add some tests
2) It would be good to have an "AAE" mode that runs every once in a while which makes sure all the neighbor and routes are matching as we expect them to be (non-delta config application)

**Comment by sargun** [comment]
otherwise, looks good.

File: apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl [hunk 3]
```diff
@@ -18,22 +18,24 @@
...
 -define(KILL_TIMER, 300000). %% 5 min
--define(LISTEN_TIMEOUT, 5000). %% 5 secs
+-define(BATCH_TIMEOUT, 5000). %% 5 secs
+-define(REAPPLY_TIMEOUT, 300000). %% 5 min
 
 -include_lib("stdlib/include/ms_transform.hrl").
 
--type config() :: #{OverlayKey :: term() := Value :: term()}.
--type config2() :: #{key := term(), value := term()}.
+-type orddict() :: [{Key::term(), Value::term()}].
```
Comment (sargun) [inline apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl]: Any reason you're not using orddict:ordict(Type1, Type1)?

File: apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl [hunk 4]
```diff
@@ -18,22 +18,24 @@
...
+-type config() :: orddict().
```
Comment (sargun) [inline apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl]: You should probably use orddict:orddict and specify the key and val types

File: apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl [hunk 5]
```diff
@@ -114,68 +116,121 @@ terminate(_Reason, _State, _Data) ->
...
--spec(update_config(config2(), OldConfig :: config()) -> config()).
-update_config(#{key := OverlayKey, value := DeltaValue}, OldConfig) ->
-    NewValue = case maps:get(OverlayKey, OldConfig, []) of
-                   [] ->
-                      ordsets:from_list(DeltaValue);
-                   OldValue ->
-                      ordsets:union(ordsets:from_list(DeltaValue), OldValue)
-               end,
-    maps:put(OverlayKey, NewValue, OldConfig).
+-spec(handle_event(LashupEvent :: tuple(), #data{}) -> #data{}).
```
Comment (sargun) [inline apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl]: Don't use handle_event, it's a special function name when gen_statem is in a different callback mode.

File: apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl [hunk 6]
```diff
@@ -114,68 +116,121 @@ terminate(_Reason, _State, _Data) ->
...
+handle_event({lashup_kv_events, #{type := ingest_new, key := OverlayKey, value := NewValue, ref := Ref}},
```
Comment (sargun) [inline apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl]: Can we come up with a better name than "Value" Maybe OverlayConfiguration?

File: rebar.config [hunk 7]
```diff
@@ -18,10 +18,11 @@
     {mesos_state, {git, "https://github.com/dcos/mesos_state.git", {branch, "master"}}},
     {enacl, {git, "https://github.com/jlouis/enacl.git", {branch, "master"}}},
     {erlzk, {git, "https://github.com/huaban/erlzk.git", {tag, "v0.6.2"}}},
-    {gpb, {git, "https://github.com/tomas-abrahamsson/gpb.git", {tag, "3.21.3"}}}
+    {gpb, {git, "https://github.com/tomas-abrahamsson/gpb.git", {tag, "3.21.3"}}},
+    {gen_netlink, ".*", {git, "https://github.com/mesosphere/gen_netlink.git", {branch, "master"}}}
 ]}.
 
-{relx, [{release, {'navstar', "0.1.0"}, [navstar, mnesia, eper, runtime_tools, eflame, debugger, observer]},
+{relx, [{release, {'navstar', "0.1.0"}, [navstar, mnesia, eper, runtime_tools, eflame, debugger, observer, {gen_netlink, load}]},
```
Comment (sargun) [inline rebar.config]: Why `{gen_netlink, load}` It's okay to start it.

**Review by sargun — CHANGES_REQUESTED** [review]

**Pull request merged**